oak_crypto/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+}
+
+/// Encryptor for client requests that will be sent to the enclave.
+pub struct ClientRequestEncryptor {
+    sender_context: SenderContext,
+    sender_response_context: SenderResponseContext,
+}
+
+impl ClientRequestEncryptor {
+    /// Encrypts `plaintext` and authenticates `associated_data` using AEAD.
+    /// Returns an encrypted request and a corresponding response decryptor.
+    /// <https://datatracker.ietf.org/doc/html/rfc5116>
+    pub fn encrypt(
+        mut self,
+        plaintext: &[u8],
+        associated_data: &[u8],
+    ) -> anyhow::Result<(AeadEncryptedMessage, ClientResponseDecryptor)> {
+        let request = self
+            .sender_context
+            .seal(plaintext, associated_data)
+            .context("couldn't encrypt request")?;
+        let decryptor = ClientResponseDecryptor {
+            sender_context: self.sender_context,
+            sender_response_context: self.sender_response_context,
+        };
+        Ok((request, decryptor))
+    }
+}
+
+/// Decryptor for enclave responses that are received by the client.
+pub struct ClientResponseDecryptor {
+    sender_context: SenderContext,
+    sender_response_context: SenderResponseContext,
+}
+
+impl ClientResponseDecryptor {
+    /// Decrypts `response.ciphertext` and authenticates `response.associated_data` using AEAD.
+    /// Returns a response message plaintext and a request encryptor for encrypting a new request
+    /// within the same session.
+    /// <https://datatracker.ietf.org/doc/html/rfc5116>
+    pub fn decrypt(
+        mut self,
+        response: &AeadEncryptedMessage,
+    ) -> anyhow::Result<(Vec<u8>, ClientRequestEncryptor)> {
+        let message = self
+            .sender_response_context
+            .open(response)
+            .context("couldn't decrypt response")?;
+        let encryptor = ClientRequestEncryptor {
+            sender_context: self.sender_context,
+            sender_response_context: self.sender_response_context,
+        };
+        Ok((message, encryptor))
+    }
+}
+
+/// Implementation of the HPKE recipient.
+/// Creates HPKE decryptors for secure bidirectional sessions with HPKE senders.
+///
+/// Each new call to the [`EnclaveCryptoProvider::create_decryptor`] creates a new decryptor using a
+/// serialized ephemeral client public key and represents a new HPKE session.
+///
+/// To prevent from reusing same encryptors and decryptors for multiple sessions, each call to the
+/// [`EnclaveRequestDecryptor::decrypt`] consumes the corresponding decryptor and produces an
+/// encryptor for the response message. And each call to [`EnclaveResponseEncryptor::encrypt`]
+/// consumes the corresponding encryptor and produces a new decryptor for decrypting a new request
+/// within the same session.
+///
+/// Public key that corresponds to the [`EnclaveCryptoProvider::key_pair`] can be used by HPKE
+/// senders to derive encryption keys for each secure bidirectional session.
+pub struct EnclaveCryptoProvider {
```

Comment:
AFAICT there is nothing enclave specific here or any of the types below?

Reply:
Changed `Client` and `Enclave` to `Sender` and `Recipient`